File: datafn/core/src/hooks.ts
/**
 * Generic plugin hook runner (HKS-001)
 * Fail-closed for before hooks, fail-open for after hooks.
 */

import type { DatafnPlugin, DatafnHookContext } from "./types.js";

export interface HookError {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

export type BeforeHookResult<T> =
  | { ok: true; value: T }
  | { ok: false; error: HookError };

type AfterHookName = "afterQuery" | "afterMutation" | "afterSync";

function wrapError(err: unknown, pluginName: string, hookName: string): HookError {
  if (err && typeof err === "object") {
    const e = err as Record<string, unknown>;
    return {
      code: (e.code as string) || "INTERNAL",
      message: (e.message as string) || "Plugin error",
      details: { path: `plugins.${pluginName}.${hookName}`, ...((e.details as Record<string, unknown>) ?? {}) },
    };
  }
  return {
    code: "INTERNAL",
    message: String(err),
    details: { path: `plugins.${pluginName}.${hookName}` },
  };
}

/**
 * Run a before-style hook across plugins filtered by env.
 * Fail-closed: first error stops the chain and returns { ok: false, error }.
 * Each hook's return value (if defined) becomes the new payload.
 *
 * @param preArgs Optional args inserted before payload in hook call: hook(ctx, ...preArgs, payload)
 */
export async function runBeforeHook<T = unknown>(
  plugins: DatafnPlugin[],
  env: "client" | "server",
  hookName: keyof DatafnPlugin,
  ctx: DatafnHookContext,
  payload: T,
  preArgs: unknown[] = [],
): Promise<BeforeHookResult<T>> {
  let current = payload;
  const filtered = plugins.filter((p) => p.runsOn.includes(env));

  for (const plugin of filtered) {
    const hook = plugin[hookName] as ((...args: unknown[]) => Promise<unknown> | unknown) | undefined;
    if (!hook) continue;

    try {
      const result = await hook(ctx, ...preArgs, current);
      if (result !== undefined) current = result as T;
    } catch (err) {
      return { ok: false, error: wrapError(err, plugin.name, String(hookName)) };
    }
  }

  return { ok: true, value: current };
}

/**
 * Run an after-style hook across plugins filtered by env.
 * Fail-open: errors are logged but do not stop the chain.
 *
 * @param preArgs Optional args inserted before payload in hook call: hook(ctx, ...preArgs, payload, result)
 */
export async function runAfterHook(
  plugins: DatafnPlugin[],
  env: "client" | "server",
  hookName: AfterHookName,
  ctx: DatafnHookContext,
  payload: unknown,
  result: unknown,
  preArgs: unknown[] = [],
): Promise<unknown> {
  const filtered = plugins.filter((p) => p.runsOn.includes(env));
  let currentResult = result;

  for (const plugin of filtered) {
    const hook = plugin[hookName] as ((...args: unknown[]) => Promise<unknown> | unknown) | undefined;
    if (!hook) continue;

    try {
      const nextResult = await hook(ctx, ...preArgs, payload, currentResult);
      if (nextResult !== undefined) {
        currentResult = nextResult;
      }
    } catch (err) {
      console.error(`Plugin ${plugin.name}.${String(hookName)} error:`, err);
    }
  }

  return currentResult;
}

File: datafn/python/datafn/hooks.py
from typing import Any, Dict, List, Optional

def _normalize_runs_on(plugin: Any) -> List[str]:
    runs_on = getattr(plugin, "runs_on", [])
    if isinstance(runs_on, list):
        return runs_on
    if isinstance(runs_on, tuple):
        return list(runs_on)
    return []

async def run_before_hook(
    plugins: List[Any],
    env: str,
    hook_name: str,
    ctx: Any,
    payload: Any,
    logger: Any = None,
) -> Dict[str, Any]:
    current = payload
    for plugin in plugins:
        runs_on = _normalize_runs_on(plugin)
        if env not in runs_on:
            continue

        hook_fn = getattr(plugin, hook_name, None)
        if hook_fn is None:
            continue

        try:
            result = await hook_fn(ctx, current)
            if result is not None:
                current = result
        except Exception as e:
            plugin_name = getattr(plugin, "name", "unknown")
            if logger:
                logger.error(
                    f"Plugin '{plugin_name}' {hook_name} failed",
                    {"plugin": plugin_name, "hook": hook_name, "error": str(e)},
                )
            return {
                "ok": False,
                "error": {
                    "code": "PLUGIN_ERROR",
                    "message": f"Plugin '{plugin_name}' {hook_name} failed",
                    "details": {"plugin": plugin_name, "hook": hook_name},
                },
            }

    return {"ok": True, "value": current}


async def run_after_hook(
    plugins: List[Any],
    env: str,
    hook_name: str,
    ctx: Any,
    payload: Any,
    result: Any,
    logger: Any = None,
) -> Any:
    current_result = result
    for plugin in plugins:
        runs_on = _normalize_runs_on(plugin)
        if env not in runs_on:
            continue

        hook_fn = getattr(plugin, hook_name, None)
        if hook_fn is None:
            continue

        try:
            transformed = await hook_fn(ctx, payload, current_result)
            if transformed is not None:
                current_result = transformed
        except Exception as e:
            if logger:
                logger.error(
                    f"Plugin '{getattr(plugin, 'name', 'unknown')}' {hook_name} failed: {e}",
                    {"plugin": getattr(plugin, "name", "unknown"), "hook": hook_name},
                )

    return current_result
